GitUI/Plugin/PluginRegistry.cs

```diff
@@ -86,9 +95,63 @@ public static void Unregister(IGitUICommands gitUiCommands)
             lock (Plugins)
             {
                 Plugins.ForEach(p => p.Unregister(gitUiCommands));
+
+                gitUiCommands.PostCommit -= OnPostCommit;
+                gitUiCommands.PostRepositoryChanged -= OnPostRepositoryChanged;
+                gitUiCommands.PostSettings -= OnPostSettings;
+                gitUiCommands.PostUpdateSubmodules -= OnPostUpdateSubmodules;
+                gitUiCommands.PostBrowseInitialize -= OnPostBrowseInitialize;
+                gitUiCommands.PostRegisterPlugin -= OnPostRegisterPlugin;
+                gitUiCommands.PreCommit -= OnPreCommit;
             }
 
             PluginsRegistered = false;
         }
+
+        #region Events
+
+        private static void OnPostCommit(object sender, GitUIPostActionEventArgs e)
+        {
+            Plugins.OfType<IPostCommitHandler>()
+                .ForEach(x => x.OnPostCommit(e));
+        }
```

Comment:
I still can't see a good reason to introduce a bunch of new interfaces. 
I can certainly see benefits in removing subscriptions from plugin implementations, and allowing them to plug into the extensions point via overrides, but we don't need to define multiple interfaces for this.

```suggestion
        private static void OnPostCommit(object sender, GitUIPostActionEventArgs e)
        {
            foreach (IPlugin plugin in Plugins)
                plugin.OnPostCommit(e);
        }
```

Reply:
- SOLID
- removing subscriptions from plugin and easier handling
- less opportunity to make a mistake
- no need to have an empty implementation for the method used

First version of implementation

![image](https://user-images.githubusercontent.com/3169265/98448281-a037eb00-2133-11eb-8e93-923ce10cbcef.png)

And then we add some new feature

![image](https://user-images.githubusercontent.com/3169265/98448268-8bf3ee00-2133-11eb-8b44-66970f4f54be.png)

The first plugin works as it did.
The second implements the new function.
The main code hasn't changed.

Reply:
> * SOLID

I am a proponent of SOLID, but I don't see how it applies here. This is an extension point by which the main app interacts with plugins. SOLID is not just SR, it is other things that are completely inappropriate here. The main app isn't loading different plugins based on interfaces, it loads all plugins.
The extensibility surface includes operations "register" and "unregister", as well other events that plugins may be interested in reacting to. It is up to each plugin author to plug into which ever flow they feel necessary.

Multiple types (i.e. interfaces) have associated costs to them - both at the runtime level (types need to loaded into memory, type resolution, etc.), and more importantly at author level - instead of dealing with a single type which has few defined and documented extension points, an plugin author must _know_ to implement specific implementations. There's too much overhead both from the project to maintain, and from an author to deal with.

Reply:
> I am a proponent of SOLID, but I don't see how it applies here.

I agree.

> SOLID is not just SR, it is other things that are completely inappropriate here.

I agree. Could you please clarify what specific things are you talking about?

> Multiple types (i.e. interfaces) have associated costs to them - both at the runtime level (types need to loaded into memory, type resolution, etc.), and more importantly at author level - instead of dealing with a single type which has few defined and documented extension points, an plugin author must know to implement specific implementations. There's too much overhead both from the project to maintain, and from an author to deal with.

I agree. Magic, but I think plugins will run faster.

Reply:
Changed to use "foreach".